src/bp/ui-shared/src/TreeView/index.tsx

```diff
@@ -0,0 +1,288 @@
+  const [nodes, setNodes] = useState<TreeNode<T>[]>([])
+  const [expanded, setExpanded] = useState(props.expandedPaths || [])
+  const [, forceUpdate] = useReducer(x => x + 1, 0)
+
+  const { elements, filter, nodeRenderer, folderRenderer, postProcessing } = props
+
+  useEffect(() => {
+    if (!elements || (elements && !nodeRenderer)) {
+      return
+    }
+
+    const nodes = buildTree({ elements, nodeRenderer, filter, folderRenderer, postProcessing })
+    traverseTree(nodes, node => {
+      if (props.visibleElements?.find(x => node.nodeData?.[x.field] === x.value)) {
+        handleNodeExpansion(node.parent!, true)
+        node.parent!.isExpanded = true
+        node.isSelected = true
+      }
+
+      if (filter?.text) {
+        node.isExpanded = true
+      }
+    })
+
+    setNodes(nodes)
+  }, [elements, filter])
+
+  useEffect(() => {
+    props.nodes && setNodes(props.nodes)
+  }, [props.nodes])
+
+  useEffect(() => {
+    setExpanded(props.expandedPaths || [])
+  }, [props.expandedPaths])
+
+  const handleNodeExpansion = (node: TreeNode<T>, isExpanded: boolean) => {
+    isExpanded ? setExpanded([...expanded, node.fullPath]) : setExpanded(expanded.filter(x => x !== node.fullPath))
+    node.isExpanded = isExpanded
+
+    props.onExpandToggle?.(node, isExpanded)
+  }
+
+  const handleNodeClick = (selectedNode: TreeNode<T>) => {
+    if (selectedNode.nodeData) {
+      props.onClick?.(selectedNode.nodeData)
+    }
+
+    traverseTree(nodes, node => {
+      if (node === selectedNode) {
+        if (props.highlightFolders || (!props.highlightFolders && node.type !== 'folder')) {
+          node.isSelected = !node.isSelected
+        }
+
+        if (!node.nodeData) {
+          node.isExpanded ? handleNodeExpansion(node, false) : handleNodeExpansion(node, true)
+        }
+      } else {
+        node.isSelected = false
+      }
+    })
+
+    forceUpdate()
+  }
+
+  const handleNodeDoubleClick = (selectedNode: TreeNode<T>) => {
+    if (selectedNode.nodeData) {
+      props.onDoubleClick?.(selectedNode.nodeData)
+    }
+  }
+
+  const handleContextMenu = (node: TreeNode<T>, _path, e) => {
+    if (!node.nodeData) {
+      return
+    }
+
+    e.preventDefault()
+
+    const contextMenu = props.onContextMenu?.(node.nodeData)
+    if (!contextMenu) {
+      return
+    }
+
+    ContextMenu.show(contextMenu, { left: e.clientX, top: e.clientY })
+  }
+
+  if (!nodes) {
+    return null
+  }
+
+  return (
+    <Tree
+      contents={nodes}
+      onNodeClick={node => handleNodeClick(node as TreeNode<T>)}
+      onNodeDoubleClick={node => handleNodeDoubleClick(node as TreeNode<T>)}
+      onNodeCollapse={node => handleNodeExpansion(node as TreeNode<T>, false)}
+      onNodeExpand={node => handleNodeExpansion(node as TreeNode<T>, true)}
+      onNodeContextMenu={(node, _p, e) => handleContextMenu(node as TreeNode<T>, _p, e)}
+      className={Classes.ELEVATION_0}
+    />
+  )
+}
+
+function splitPath<T>(
```

Comment:
Could these functions be in a utils file in the TreeView folder to declutter the file?

Reply:
Good point